lib/Validation.ml

```diff
@@ -0,0 +1,55 @@
+include Validation_intf
+include Core_kernel
+
+module Make (X : sig
+  type t
+end) : S with type error := X.t = struct
+  type errors = NonEmpty of X.t * X.t list
+
+  let append xs ys =
+    match (xs, ys) with
+    | NonEmpty (x, []), NonEmpty (y, []) -> NonEmpty (x, [y])
+    | NonEmpty (x, xs), NonEmpty (y, ys) -> NonEmpty (x, xs @ (y :: ys))
+
+  type 'a t = ('a, errors) result
+
+  let map x ~f = match x with Ok x -> Ok (f x) | Error x -> Error x
+  let pure x = Ok x
+
+  let apply x ~f =
+    match (f, x) with
+    | Ok f, Ok x -> Ok (f x)
+    | Error e, Ok _ -> Error e
+    | Ok _, Error e -> Error e
+    | Error e1, Error e2 -> Error (append e1 e2)
+
+  let apply_const a b = apply b ~f:(map a ~f:(fun _ x -> x))
+  let liftA2 f x y = apply y ~f:(apply x ~f:(pure f))
+  let consA next rest = liftA2 List.cons next rest
+  let sequence ts = List.fold_right ~init:(pure []) ~f:consA ts
+
+  module Validation_infix = struct
+    let ( <*> ) x f = apply x ~f
+    let ( *> ) x y = apply_const x y
```

Comment:
Let's stick to the 4 monadic punctuations that they use everywhere in Core_kernel? `|>`, `@@`, `>>=` and `>>|` and use names for everything else if you don't mind.

Reply:
Sure - I do agree the infix operators don't buy us a huge amount and will look cryptic unless you happen to know what an applicative functor is...